src/App.tsx

```diff
@@ -1,61 +1,163 @@
+  const getUserById = (userId: number): User | null => {
+    return usersFromServer.find(user => user.id === userId) || null;
+  };
+
+  // Initial processing of todos to include the user object
+  const prepareInitialTodos = (): Todo[] => {
+    return todosFromServer.map(todo => ({
+      ...todo,
+      user: getUserById(todo.userId),
+    }));
+  };
+
+  // State initialization
+  const [userList] = useState(usersFromServer);
+  const [todosList, setTodosList] = useState(prepareInitialTodos());
+  const [titleInput, setTitleInput] = useState('');
+  const [userSelected, setUserSelected] = useState('');
+  const [titleInputError, setTitleInputError] = useState(false);
+  const [userSelectError, setUserSelectError] = useState(false);
+
+  // Helper function to get the next available ID
+  const getNextTodoId = (): number => {
+    const allTodoIds = todosList.map(todo => todo.id);
+
+    return Math.max(...allTodoIds, 0) + 1;
+  };
+
+  // Create a new Todo object
+  const createNewTodo = (): Todo => {
+    const trimmedTitle = titleInput.trim();
+
+    return {
+      id: getNextTodoId(),
+      title: trimmedTitle,
+      completed: false,
+      userId: Number(userSelected),
+      user: getUserById(Number(userSelected)),
+    };
+  };
+
+  // Form input validators
+  const validateTitleInput = () => {
+    const isTitleEmpty = titleInput.trim() === '';
+
+    setTitleInputError(isTitleEmpty);
+
+    return !isTitleEmpty;
+  };
+
+  const validateUserSelection = () => {
+    const isUserUnselected = userSelected === '';
+
+    setUserSelectError(isUserUnselected);
+
+    return !isUserUnselected;
+  };
+
+  // Reset form fields
+  const resetForm = () => {
+    setTitleInput('');
+    setUserSelected('');
+    setTitleInputError(false);
+    setUserSelectError(false);
+  };
+
+  // Form submission handler
+  const handleFormSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+
+    const isTitleValid = validateTitleInput();
+    const isUserValid = validateUserSelection();
+
+    if (isTitleValid && isUserValid) {
+      const newTodo = createNewTodo();
+
+      setTodosList(prevTodos => [...prevTodos, newTodo]);
+
+      resetForm();
+    }
+  };
+
+  // Input change handlers
+  const handleTitleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const sanitizedValue = event.target.value.replace(
+      /[^a-zA-Zа-яА-Я0-9 ]/g,
+      '',
+    );
+
+    setTitleInput(sanitizedValue);
+
+    if (sanitizedValue.trim() !== '') {
+      setTitleInputError(false);
+    }
+  };
+
+  const handleUserChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    setUserSelected(event.target.value);
+
+    if (event.target.value !== '') {
+      setUserSelectError(false);
+    }
+  };
+
   return (
     <div className="App">
-      <h1>Add todo form</h1>
+      <h1>Add Todo Form</h1>
 
-      <form action="/api/todos" method="POST">
+      <form action="/api/todos" method="POST" onSubmit={handleFormSubmit}>
```

Comment:
The `action` attribute in the form tag is set to `/api/todos`. If the form submission is handled entirely on the client side, this attribute might not be necessary and could be removed to avoid confusion.